plan/dag_plan_test.go

```diff
@@ -120,19 +138,78 @@ func (s *testPlanSuite) TestDAGPlanBuilderSimpleCase(c *C) {
 		stmt, err := s.ParseOneStmt(tt.sql, "", "")
 		c.Assert(err, IsNil, comment)
 
-		is, err := mockResolve(stmt)
+		is, err := plan.MockResolve(stmt)
 		c.Assert(err, IsNil)
 
-		builder := &planBuilder{
-			allocator: new(idAllocator),
-			ctx:       mockContext(),
-			colMapper: make(map[*ast.ColumnNameExpr]int),
-			is:        is,
-		}
-		p := builder.build(stmt)
-		c.Assert(builder.err, IsNil)
-		p, err = doOptimize(builder.optFlag, p.(LogicalPlan), builder.ctx, builder.allocator)
 		c.Assert(err, IsNil)
-		c.Assert(ToString(p), Equals, tt.best, Commentf("for %s", tt.sql))
+		p, err := plan.Optimize(se, stmt, is)
+		c.Assert(err, IsNil)
+		c.Assert(plan.ToString(p), Equals, tt.best, Commentf("for %s", tt.sql))
+	}
+}
+
+func (s *testPlanSuite) TestDAGPlanBuilderUnionScan(c *C) {
+	store, err := newStoreWithBootstrap()
+	c.Assert(err, IsNil)
+	defer store.Close()
+	se, err := tidb.CreateSession(store)
+	c.Assert(err, IsNil)
+
+	plan.UseDAGPlanBuilder = true
+	defer func() {
+		plan.UseDAGPlanBuilder = false
+		testleak.AfterTest(c)()
+	}()
+	tests := []struct {
+		sql  string
+		best string
+	}{
+		// Read table.
+		{
+			sql:  "select * from t",
+			best: "TableReader(Table(t))->UnionScan",
+		},
+		{
+			sql:  "select * from t where b = 1",
+			best: "TableReader(Table(t)->Sel([eq(test.t.b, 1)]))->UnionScan(eq(test.t.b, 1))",
+		},
+		{
+			sql:  "select * from t where a = 1",
+			best: "TableReader(Table(t))->UnionScan(eq(test.t.a, 1))",
+		},
+		{
+			sql:  "select * from t where a = 1 order by a",
+			best: "TableReader(Table(t))->UnionScan(eq(test.t.a, 1))",
+		},
+		{
+			sql:  "select * from t where a = 1 order by b",
+			best: "TableReader(Table(t))->UnionScan(eq(test.t.a, 1))->Sort",
+		},
+		{
+			sql:  "select * from t where a = 1 limit 1",
+			best: "TableReader(Table(t))->UnionScan(eq(test.t.a, 1))->Limit",
+		},
+		{
+			sql:  "select * from t where c = 1",
+			best: "IndexLookUp(Index(t.c_d_e)[[1,1]], Table(t))->UnionScan(eq(test.t.c, 1))",
+		},
+		{
+			sql:  "select c from t where c = 1",
+			best: "IndexReader(Index(t.c_d_e)[[1,1]])->UnionScan(eq(test.t.c, 1))",
+		},
+	}
+	for _, tt := range tests {
+		comment := Commentf("for %s", tt.sql)
+		stmt, err := s.ParseOneStmt(tt.sql, "", "")
+		c.Assert(err, IsNil, comment)
+
+		is, err := plan.MockResolve(stmt)
+		c.Assert(err, IsNil)
+
+		err = se.NewTxn()
```

Comment:
New transaction without any data written doesn't need union scan.
We should test union scan with and without dirty data.